app/api/business/brief_overview_business.py

```diff
@@ -1,8 +1,76 @@
 from flask_login import current_user
+
 from app.api.business.agreement_business import use_old_work_order_creator
+from app.api.business.errors import (BriefError, NotFoundError,
+                                     UnauthorisedError)
+from app.api.services import (audit_service, audit_types,
+                              brief_responses_service, briefs)
+from app.emails import send_opportunity_closed_early_email
+from app.tasks import publish_tasks
+from app.tasks.s3 import create_responses_zip
 from app.validation import get_sections as get_validation_sections
 
 
+def can_close_opportunity_early(brief):
+    seller_selector = brief.data.get('sellerSelector', '')
+    invited_sellers = brief.data.get('sellers', {}).keys()
+    number_of_candidates = int(brief.data.get('numberOfSuppliers', 0))
+
+    if brief.status == 'live' and len(invited_sellers) == 1 and (
+        (brief.lot.slug in ['rfx', 'training2'] and seller_selector == 'oneSeller') or
+        (brief.lot.slug == 'specialist' and seller_selector == 'someSellers')
+    ):
+        invited_seller_code = int(invited_sellers.pop())
+        response_count = brief_responses_service.find(
+            brief_id=brief.id,
+            supplier_code=invited_seller_code,
+            status='submitted'
+        ).count()
+
+        if brief.lot.slug in ['rfx', 'training2'] and response_count == 1:
+            return True
+
+        if brief.lot.slug == 'specialist' and response_count == number_of_candidates:
+            return True
+
+    return False
+
+
+def close_opportunity_early(current_user, brief_id):
+    brief = briefs.get(brief_id)
+    if not brief:
+        raise NotFoundError('Opportunity {} does not exist'.format(brief_id))
+
+    if not briefs.has_permission_to_brief(current_user.id, brief_id):
+        raise UnauthorisedError('Not authorised to close opportunity {}'.format(brief_id))
+
+    if not can_close_opportunity_early(brief):
+        raise BriefError('Unable to close opportunity {}'.format(brief_id))
+
+    brief = briefs.close_opportunity_early(brief)
+
+    audit_service.log_audit_event(
+        audit_type=audit_types.close_opportunity_early,
+        data={
+            'briefId': brief.id
+        },
+        db_object=brief,
+        user=current_user.email_address
+    )
+
+    create_responses_zip(brief.id)
```

Comment:
Was calling this as an async `.delay()` considered? I don't know if it is necessary as this would only fire when there is one response, but given one response can have a few attachments (up to 10), and they can all be big enough to be a few hundred megabytes in total, it is possible this call may be quite slow. Slow enough to risk timing out for the user I doubt, but it might be possible. On the other hand having this async would mean it is possible a buyer beats the async task to the punch and gets confused by the download not being available on their overview page (which is not built for the possibility the download is being processed but not ready yet).

Reply:
These are all great points.  If we make this async, we would probably need to rethink the user experience and what the design looks like.  Lets see how this performs and we can make improvements if we need to.